### ISSUE TITLE
MemberNotNullWhen seems ignored with generic types

### ISSUE THREAD

**Issue by meziantou** [dotnet/roslyn#47915]
**Version Used**: `'3.8.0-3.20458.6 (b446afd3)'. Language version: 9.0`

**Steps to Reproduce**:

The following types are similar. The second one is the generic version of the first one (with class constraint).

````c#
    public sealed class NonGeneric
    {
        public NonGeneric(string? value) => Value = value;

        [MemberNotNullWhen(returnValue: true, member: "Value")]
        public bool IsNotNull => Value != null;

        public string? Value { get; }
    }

    public sealed class Generic<T> where T : class
    {
        public Generic(T? value) => Value = value;

        [MemberNotNullWhen(returnValue: true, member: "Value")]
        public bool IsNotNull => Value != null;

        public T? Value { get; }
    }
````

In the case of the generic class, the `MemberNotNullWhen` seems ignored

````c#
            var demo = new NonGeneric("");
            if (demo.IsNotNull)
            {
                _ = demo.Value.Length; // Ok, no warning
            }

            var demo2 = new Generic<string>("");
            if (demo2.IsNotNull)
            {
                _ = demo2.Value.Length; // CS8602, but I think this is not expected
            }
````

**Expected Behavior**:

The compiler doesn't report CS8602 when using `demo2.Value.Length`

**Actual Behavior**:

The compiler reports CS8602 when using `demo2.Value.Length`

**Comment by jcouv** [comment]
I was not able to reproduce this. I see no nullability warnings on either statements in [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAobABgAJt8A6AEQEsIBzAOwHtYGBjKGgYXYo4AQVYQANgE8oDKAG4SJAMSsArhIkRgEuBTjidcEgDcICCsLDsKAXgqs4AdwoA5dqwDiBxDwAUAEQBAJSKxAwAZhR+Vuw0AJJQ7jCuGhLBFADeJBQUAPp2lnDWNABqkmpwNAAyBswwABbyFAD0LRQA8gDWSA42TuasDKzMJAC+SsRmFrG4hY4u3o4IPBQAPFT4lAB80UGhJJHRswlJ7ClpGdnEuQX2J+USlTV1jc1tFPwAygAcAGxkXC9YBqGAUeIURrDLqQhpyCjwjhguAADwADnBuPAUONJtgAMwUKBwSRwFBUObuLw+FbcLI5KiEqlLXzcPzUSgAfgoZiecAytm2j0qhV5lTCDIA2gBZYrARDJVKaADqDQM7IA7MK4CBIQhKr0wHLELqAtqQgBdBkEijAdjsCTgs4XTR2IUVPQAQns6k0EpujM2XIo2vpAdyzDgMDCuQmxDjJBtxNJ5LwFBZtPWFAAKhRdk41Qg9NndWnrrkbRn/LnuWL+W7Q/Y6/7cjLjQhFWlVersFqPbqYPq4Ib26bzcErQGbXaHU7O67BaHvQ40i3AzWQx6w7kI1GYxQ4wniB9qjwDMTyTAbI09DRXABRXMAMXYalYKAgMAYHgob+EFg8PR2AsawiwoFgi2KAwYD4EgPmzNUKDvR8KBfP9P2/VgKAkM9WGJKBYXhCIGF0SEbCkV9f3fRBYT0aV4lzHDuHPKpJnEI0oDRCBmM2egmDYTgv14AQhFEcRpFkAjrilEQYEHBgQXgABVKAWDgPxZPkxS4GzcxIxgmhZUaIQKEwChNJWbTdIQfS+AABQQdgMQQGApF6eJWELBhsUKCJJGJXoRE0dgnGlDQvzRUj7EHSoJ0mCtCWTXRUzmWUwHlDtziVCRu1YCyFNBHVzLkyzCu3HcbTSjL5xytU8pKgr4D8GdHV7bVeg5CgjXSxArgZHcKAAJSjNQEFYRsqD7Pl9wGqrEAIn1nElC1yoG3Juoy/qdzGGaD3iirCTmzKXVqgx8u05r7VaqaDQoLiEAgMACI5ZauvbKA+vDHdhpgUbxq3ew2o9XbciOha3p6hAFC2o8BunK6hpGsbQ3LNb9N22GDqDV6wdWnd0ZhhlDyAA==).

I think this was fixed in 16.8p4 by PR https://github.com/dotnet/roslyn/pull/47818 (MemberNotNullWhen not taking effect when type substitutions are involved).

I'll go ahead and close the issue.
Thanks for reporting.

**Comment by PovilasAtPoq** [comment]
@jcouv can you tell in which version it was fixed? I have identical issue, and I'm running c# 9.0 and building with .NET SDK 5.0.401 (x64) (5.4.121.42430)

**Comment by jcouv** [comment]
@PovilasAtPoq Sorry, I don't have an easy way to tell which version of the SDK this compiler fix flowed into. You can put `#error version` in your source code to see the compiler version that you're using (it'll show a version number and a git SHA as a diagnostic). Looking at the SHA on the roslyn repo, we can tell whether it is before or after the fix I mentioned above.
